src/main/java/com/commercetools/sync/sdk2/types/utils/FieldDefinitionsUpdateActionUtils.java

```diff
@@ -0,0 +1,288 @@
+package com.commercetools.sync.sdk2.types.utils;
+
+import static com.commercetools.sync.sdk2.commons.utils.CommonTypeUpdateActionUtils.buildUpdateAction;
+import static com.commercetools.sync.sdk2.types.utils.FieldDefinitionUpdateActionUtils.buildActions;
+import static java.lang.String.format;
+import static java.util.Collections.singletonList;
+import static java.util.Optional.ofNullable;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toMap;
+
+import com.commercetools.api.models.type.FieldDefinition;
+import com.commercetools.api.models.type.FieldDefinitionBuilder;
+import com.commercetools.api.models.type.FieldType;
+import com.commercetools.api.models.type.TypeAddFieldDefinitionActionBuilder;
+import com.commercetools.api.models.type.TypeChangeFieldDefinitionOrderActionBuilder;
+import com.commercetools.api.models.type.TypeRemoveFieldDefinitionActionBuilder;
+import com.commercetools.api.models.type.TypeUpdateAction;
+import com.commercetools.sync.sdk2.commons.exceptions.BuildUpdateActionException;
+import com.commercetools.sync.sdk2.commons.exceptions.DuplicateKeyException;
+import com.commercetools.sync.sdk2.commons.exceptions.DuplicateNameException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+/** This class is only meant for the internal use of the commercetools-sync-java library. */
+final class FieldDefinitionsUpdateActionUtils {
+
+  /**
+   * Compares a list of {@link FieldDefinition}s with a list of {@link FieldDefinition}s. The method
+   * serves as an implementation for field definitions syncing and building the required update
+   * actions (AddFieldDefinition, RemoveFieldDefinition, ChangeFieldDefinitionOrder) and 1-1 update
+   * actions on field definitions (e.g. changeFieldDefinitionLabel, etc..) for the required
+   * resource.
+   *
+   * <p>If the list of new {@link FieldDefinition}s is {@code null}, then remove actions are built
+   * for every existing field definition in the {@code oldFieldDefinitions} list.
+   *
+   * <p>Note: The method will ignore/filter out {@code null} field definitions from the passed
+   * {@code newFieldDefinitions}.
```

Comment:
This is a bit difficult... maybe you can rephrase to make it easier.

Reply:
https://github.com/commercetools/commercetools-sync-java/pull/1007/commits/278657792bb36fc639974c02859d071b19e02b41